pkg/backend/display/rows.go

```diff
@@ -305,17 +305,41 @@ func (data *resourceRowData) ColorizedColumns() []string {
 
 	diagInfo := data.diagInfo
 
-	if done {
-		failed := data.failed || diagInfo.ErrorCount > 0
-		columns[statusColumn] = data.display.getStepDoneDescription(step, failed)
-	} else {
-		columns[statusColumn] = data.display.getStepInProgressDescription(step)
-	}
+	failed := data.failed || diagInfo.ErrorCount > 0
 
+	columns[statusColumn] = data.display.getStepStatus(step, done, failed)
 	columns[infoColumn] = data.getInfoColumn()
 	return columns
 }
 
+// addRetainStatusFlag adds a "[retain]" suffix to the input string if the resource is marked as
+// RetainOnDelete and the step will discard the resource.
+func addRetainStatusFlag(status string, step engine.StepEventMetadata) string {
+	// Guard against nil pointer dereference and return the status string as is.
+	if step.Old == nil {
+		return status
+	}
+
+	// Guard against nil pointer dereference and return the status string as is.
+	if step.Old.State == nil {
+		return status
+	}
+
+	// If a resource is not marked as RetainOnDelete, return the status string as is.
+	if !step.Old.State.RetainOnDelete {
+		return status
```

Comment:
It's fine for this return to be separate from the "guard against nil" above,
but again, the comment here doesn't add much value versus reading the code.
There's no additional nuance here besides the obvious: if the resource isn't marked as RetainOnDelete, then we don't need to add the marker.

We can either delete the comment or add a shorter "why":

```suggestion
	// Ignore resources not retained on delete.
	if !step.Old.State.RetainOnDelete {
		return status
```

```suggestion
	if !step.Old.State.RetainOnDelete {
		return status
```